cedar-policy-core/src/evaluator.rs

```diff
@@ -422,15 +422,44 @@ impl<'e> Evaluator<'e> {
-                            (ValueKind::ExtensionValue(x), ValueKind::ExtensionValue(y)) if x.typename() == y.typename() => Err(EvaluationError::type_error_with_advice(Extensions::types_with_operator_overloading().map(|name| Type::Extension { name} ), &arg1, "Only extension types `datetime` and `duration` support operator overloading".to_string())),
+                            (ValueKind::Lit(Literal::Long(_)), _) => {
+                                Err(EvaluationError::type_error_single(Type::Long, &arg2))
+                            }
+                            (_, ValueKind::Lit(Literal::Long(_))) => {
+                                Err(EvaluationError::type_error_single(Type::Long, &arg1))
+                            }
+                            (ValueKind::ExtensionValue(x), _)
+                                if x.supports_operator_overloading() =>
+                            {
+                                Err(EvaluationError::type_error_single(
+                                    Type::Extension { name: x.typename() },
+                                    &arg2,
+                                ))
+                            }
+                            (_, ValueKind::ExtensionValue(y))
+                                if y.supports_operator_overloading() =>
+                            {
+                                Err(EvaluationError::type_error_single(
+                                    Type::Extension { name: y.typename() },
+                                    &arg1,
+                                ))
+                            }
                             _ => {
-                                let mut expected_types = Extensions::types_with_operator_overloading().map(|name| Type::Extension { name });
-                                expected_types.push(Type::Long);
-                                Err(EvaluationError::type_error_with_advice(expected_types, &arg1, "Only `Long` and extension types `datetime`, `duration` support comparison".to_string()))
+                                let mut expected_types = nonempty::NonEmpty::singleton(Type::Long);
+                                expected_types.extend(
+                                    self.extensions
+                                        .types_with_operator_overloading()
+                                        .map(|n| Type::Extension { name: n.clone() })
+                                        .collect::<BTreeSet<_>>(),
+                                );
+                                Err(EvaluationError::type_error_with_advice(
+                                    expected_types.clone(),
+                                    &arg1,
+                                    format!(
+                                        "Only type {} support comparison",
```

Comment:
```suggestion
                                        "Only types {} support comparison",
```

also, could re-use https://github.com/cedar-policy/cedar/blob/4c9c8d892982df5e540fdfe6c569249551525159/cedar-policy-core/src/parser/fmt.rs#L458